docs/design/2020-06-27-non-recursive-common-table-expression.md

```diff
@@ -0,0 +1,143 @@
+
+## Proposal
+
+We design two strategies for CTE: __Merge__ & __Materialization__:
+
+#### Merge
+The CTE is expanded on where it's referenced:
+```sql
+WITH cte(k, v) AS (SELECT i, j FROM t1)
+SELECT v FROM cte WHERE k > 10
+UNION DISTINCT
+SELECT v FROM cte WHERE k < 100
+```
+After "merged":
+```sql
+SELECT v FROM (SELECT i AS k, j AS v FROM t1) WHERE k > 10
+UNION DISTINCT
+SELECT v FROM (SELECT i AS k, j AS v FROM t1) WHERE k < 100
+```
+
+We will add some rules to rewrite the plan trees, similar to `subquery optimization` [2].
+
+_(`Merge` is also called `Inline` is some other DBMS [4][11])_
+
+#### Materialization
+A CTE is first executed and stored result temporarily during the statement execution.
+
+Then, on each point the CTE is referenced, the tuples of CTE is read from the temporary storage.
+
+To improve performance, the temporary result is stored in memory first, then spill to disk if out of memory quota, similar to disk-based executors [3].
+
+Besides, [4] introduces a `Producer-Consumer` model. We can utilize this model to make writing and reading of CTE parallel.
+
+![cte01](imgs/cte01.png)
+
+#### Optimization
+For `Merge` strategy, most existed rules for subquery can be reused.
+
+For `Materialization` strategy, disjunction of all predicates on top can be push down to CTE [4].
+
+Finally, the costs should be estimated, to select the better strategy for each CTE.
+
+#### Optimizer Hints
+To be compatible to MySQL [1], we propose these hints:
+* __MERGE__: indicates the CTE should be merged.
+* __NO_MERGE__: indicates the CTE should be materialized.
+
+E.g.
+```sql
+WITH
+  cte1 AS (SELECT a, b FROM table1),
+  cte2 AS (SELECT c, d FROM table2)
+SELECT /*+ MERGE(cte1) NO_MERGE(cte2) */ cte1.b, cte2.d
+FROM cte1 JOIN cte2
+WHERE cte1.a = cte2.c;
+```
+_(cte1 should be merged and cte2 should be materialized)_
+
+
+
+## Rationale
+
+Mainstream database systems utilize `Merge` and `Materialization` to implement CTE.
+
+In these DBMS, `Merge` is preferred when the CTE has no side-effect, and is referenced in a very few times. `Materialization` should be avoid if possible.
+
+If `Materialization` is necessary, these optimizations can be applied:
+1. Materialized in memory [13].
+2. Postpones until its contents are needed [1].
+3. Add index(es) according to access method [1].
+4. Pushdown predicates, and pushdown disjunction of all predicates if referenced multiple times [4][5].
+
+* MySQL using these two strategies: 1) Merge the derived table into the outer query block, or 2) Materialize the derived table to an internal temporary table [1]. The optimizer avoids unnecessary materialization whenever possible, which enables pushing down conditions from the outer query to derived tables and produces more efficient execution plans [1].
+
+* MariaDB using `Materialization` as basic algorithm. To optimize performance, MariaDB using `Merge` strategy, and utilizes "Condition Pushdown" [5].
```

Comment:
Could you describe more on how MariaDB combines these two optimizations and what can be adopted in TiDB?

Reply:
Similar to MySQL, to optimize performance, MariaDB will always try to use `Merge` strategy, and fall back to materialization due to side-effect [[18]](https://github.com/MariaDB/server/blob/5b8ab1934a10966336e66751bc13fc66923b02f6/sql/table.cc#L9224).

_([design doc](https://github.com/pingyu/tidb/blob/executor_non_recursive_cte/docs/design/2020-06-27-non-recursive-common-table-expression.md) is updated too. PTAL)_